File: app/src/main/java/com/emptycup/cupmvp/basednxymvpstudents/demo/support/ProxyMvpCallBack.java
package com.emptycup.cupmvp.basednxymvpstudents.demo.support;

import com.emptycup.cupmvp.basednxymvpstudents.demo.MVPView;
import com.emptycup.cupmvp.basednxymvpstudents.demo.MvpPresenter;

/**
 * Created by Administrator on 2018/2/6.
 * 第二重代理   代理对象
 */

public class ProxyMvpCallBack<V extends MVPView,P extends MvpPresenter<V>> implements MvpCallBack<V,P> {
    private MvpCallBack<V,P> mvpCallBack;

    public ProxyMvpCallBack(MvpCallBack<V,P> mvpCallBack){
        this.mvpCallBack = mvpCallBack;
    }
    @Override
    public P createPresnter() {
        //创建P层
        P presenter = mvpCallBack.getPresenter();
        if (presenter == null){
            presenter = mvpCallBack.createPresnter();
        }
        if (presenter == null){
            throw new NullPointerException("presenter Is Null");
        }
        this.mvpCallBack.setPresenter(presenter);
        return presenter;
    }

    @Override
    public V createView() {
        V view = mvpCallBack.getMvpView();
        if (view == null){
            view = mvpCallBack.createView();
        }
        if (view == null){
            throw  new NullPointerException("view Is Null");
        }
        return view;
    }

    @Override
    public P getPresenter() {
        return mvpCallBack.getPresenter();
    }

    @Override
    public void setPresenter(P presenter) {
        mvpCallBack.setPresenter(presenter);
    }

    @Override
    public V getMvpView() {
        return mvpCallBack.getMvpView();
    }
    public void attachView(){
        getPresenter().attachView(getMvpView());
    }
    public void detachView(){
        getPresenter().detachView();
    }
}

File: app/src/main/java/com/emptycup/cupmvp/kotlinmvp/support/ProxyMvpCallBack.kt
package com.emptycup.cupmvp.kotlinmvp.support

import android.util.Log
import com.emptycup.cupmvp.kotlinmvp.basemvp.IBaseMvpPresenter
import com.emptycup.cupmvp.kotlinmvp.basemvp.IBaseMvpView

/**
 * Created by Administrator on 2018/1/9.
 * 第二次代理的 目标实现
 */
class ProxyMvpCallBack<V: IBaseMvpView,  P: IBaseMvpPresenter<V>>(var mvpCallBack: MvpCallBack<V, P>) : MvpCallBack<V, P> {

    override fun createPresenter(): P {
        var presenter:P = mvpCallBack.getPresenter()
        if (presenter == null){
            presenter = mvpCallBack.createPresenter()
        }
        if (presenter == null){
            throw  NullPointerException("Presenter Is Null")
        }
        mvpCallBack.setPresenter(presenter)
        return presenter
    }

    override fun createView(): V {
        var view:V = mvpCallBack.getView()
        if (view == null){
            view = mvpCallBack.createView()
        }
        if (view == null){
            throw  NullPointerException("View Is Null")
        }
        return view
    }

    override fun getPresenter(): P {
        return mvpCallBack.getPresenter()
    }

    override fun setPresenter(presenter: P) = mvpCallBack.setPresenter(presenter)

    override fun getView(): V {
        return mvpCallBack.getView()
    }

    fun attachView(){
        Log.v("attachView","     getBaseView        "+getView())
        getPresenter().attachView(getView())
    }

    fun detachVeiw(){
        getPresenter().detachView()
    }

}